### PR TITLE
Add type completeness check to CI

### PR DESCRIPTION
**Pull request by desertaxle**
This PR adds a type completeness check to the Static Analysis workflow. The check uses `pyright` to calculate a type completeness score and ensures that a PR will not decrease the type completeness score by comparing the generated score to the score generated on the base branch. 

Contributors will be able to see the output in the summary of the Static Analysis workflow:
Screenshot 2024-12-04 at 9 33 58 AM

Inspired by discussion in https://github.com/PrefectHQ/prefect/discussions/16167

### CODE REVIEW

File: .github/workflows/static-analysis.yaml [hunk 1]
```diff
@@ -67,3 +67,52 @@ jobs:
...
+
+          if (( $(echo "$BASE_SCORE > $CURRENT_SCORE" | bc -l) )); then
+            echo "❌ Type completeness score has decreased from $BASE_SCORE to $CURRENT_SCORE" >> $GITHUB_STEP_SUMMARY
+            echo "Please add type annotations to your code to increase the type completeness score." >> $GITHUB_STEP_SUMMARY
+            exit 1
+          elif (( $(echo "$BASE_SCORE < $CURRENT_SCORE" | bc -l) )); then
+            echo "✅ Type completeness score has increased from $BASE_SCORE to $CURRENT_SCORE" >> $GITHUB_STEP_SUMMARY
+          else
+            echo "✅ Type completeness score remained unchanged at $BASE_SCORE" >> $GITHUB_STEP_SUMMARY
+          fi
```
Comment (zzstoatzz) [inline .github/workflows/static-analysis.yaml:118]: would we want something like this?
```suggestion
          fi

    - name: Upload analysis artifacts
      uses: actions/upload-artifact@v4
      if: always()  # Upload even if the check fails
      with:
        name: type-completeness-reports
        path: |
          prefect-analysis.json
          prefect-analysis-base.json
```

**Review by zzstoatzz — APPROVED** [review]
lgtm! just one question about potentially uploading artifacts

File: .github/workflows/static-analysis.yaml (hunk 1 above)
Comment (desertaxle) [reply .github/workflows/static-analysis.yaml:118]: Maybe? I could see that being useful if we wanted to look at historical data, and I'd probably only upload the current branch, but I can't think of a scenario where we'd want that. If we want the report, we can always generate it locally. 

Did you have a use case in mind that I'm missing?

Comment (zzstoatzz) [reply .github/workflows/static-analysis.yaml:118]: this was the only thing I was thinking as well

> useful if we wanted to look at historical data

Comment (desertaxle) [reply .github/workflows/static-analysis.yaml:118]: Gotcha, I think it's unlikely that we'll need that, but if we do, we can always checkout a commit and generate the `pyright` report on that commit.

File: .github/workflows/static-analysis.yaml [hunk 2]
```diff
@@ -67,3 +67,52 @@ jobs:
...
+          BASE_SCORE=$(echo ${{ steps.calculate_base_score.outputs.base_score }})
+
+          if (( $(echo "$BASE_SCORE > $CURRENT_SCORE" | bc -l) )); then
+            echo "❌ Type completeness score has decreased from $BASE_SCORE to $CURRENT_SCORE" >> $GITHUB_STEP_SUMMARY
+            echo "Please add type annotations to your code to increase the type completeness score." >> $GITHUB_STEP_SUMMARY
+            exit 1
+          elif (( $(echo "$BASE_SCORE < $CURRENT_SCORE" | bc -l) )); then
+            echo "✅ Type completeness score has increased from $BASE_SCORE to $CURRENT_SCORE" >> $GITHUB_STEP_SUMMARY
+          else
+            echo "✅ Type completeness score remained unchanged at $BASE_SCORE" >> $GITHUB_STEP_SUMMARY
```
Comment (mjpieters) [inline .github/workflows/static-analysis.yaml:117]: You probably want to tee the message to stdout too, so anyone reading the job log can _also_ see the outcome. Not everyone will be able to find their way to the workflow summary from a failed job (which is what the PR UI links to).

**Pull request merged**